src/index.ts

```diff
@@ -1,82 +1,162 @@
-import path from "path";
 import fs from "fs";
-import {Client, Intents, Collection, Interaction, CacheType} from "discord.js";
+import {CacheType, Client, Collection, Intents, Interaction} from "discord.js";
 import {LoadConfig, Config} from "./config";
 import {logger} from "./logger";
+import {BotModes, ClientType, CommandType, PermissionType} from "./types";
 import {
+  GlobalClearCommands,
   GlobalRegisterSlashCommands,
+  GuildClearCommands,
+  GuildRegisterPermissions,
   GuildRegisterSlashCommands,
 } from "./registerer";
-import {BotModes, ClientType, CommandType} from "./types";
+import {
+  ResolveCommandPath,
+  ResolveEventPath,
+  ResolveRoleId,
+} from "./helpers/resolvers";
 
 // start bot async function
 // needs to be async so we can `await` inside
 const start = async () => {
   // even though this is inside `src/`, pretend that it isnt
   // load in the config
   LoadConfig("config.yaml");
-
   logger.debug({Config});
 
-  // create client as ClientType
+  // create client as ClientType, then establish connection with Discord
   const client: ClientType = new Client({
     intents: [Intents.FLAGS.GUILDS],
   }) as ClientType;
+  client.commands = new Collection<string, CommandType>();
+
+  client.once("ready", () => logger.info("Bot has logged in."));
+  await client.login(Config.api_token);
+
+  /*
+   * Clear all registered commands
+   */
+  let promises: Promise<any>[] = [];
+  promises.push(GlobalClearCommands());
+  if (Config.guilds)
+    for (const guildId of Config.guilds?.keys())
```

Comment:
explicit scoping is better with clarity and helps if we ever have to minimize the bot. 
```suggestion
    for (const guildId of Config.guilds?.keys()){
        ...
    }
```

BTW this goes for all scoping in this project